pages/docs/protocols/masp/swap.mdx

```diff
@@ -0,0 +1,314 @@
+---
+title: MASP Swap Circuit
+description: Describes the MASP Swap Circuit.
+---
+
+import { MASPPrivateTransferImage, MASPSwapImage, MASPSwapDiagramImage }  from '../../../../components/images/masp/Masp.tsx'
+import Callout from "../../../../components/Callout";
+
+# MASP Swap Circuit
+
+## Updated Key Structure
+
+Each MASP participant has a set of keys of the following form:
+
+<MASPPrivateTransferImage />
+
+<Callout type="info">
+    Note: `(sk, ak)` form a secret/public keypair which can be used to make EdDSA signatures. These EdDSA signatures can then be verified inside a ZKP (`circomlib` provides templates for this).
+</Callout>
+
+## Updated Record Structure
+
+The MASP key is used in the commitment (record) that is inserted into the MASP’s Merkle tree.
+
+```markdown
+InnerPartialRecord = Poseidon(blinding)
+PartialRecord = Poseidon(DestinationChainId, PublicKey_X, PublicKey_Y, InnerPartialRecord)
+Record = Poseidon(AssetId, TokenId, Amount, PartialRecord)
+```
+
+## Swap User Mechanics
+
+<MASPSwapImage />
+
+1. Alice wants to exchange 1 webbBTC (`assetId = 1, tokenId = 0`) for Bob’s 10 webbETH (`assetId = 2, tokenId = 0`) as long as the swap occurs between time `t` and `tPrime`. <br /> <br />
+   **a.** Alice has a `Record` worth 1.5 webbBTC under her public key that she wants to use for the swap. This is known as Alice’s spend `Record`. <br />
+   **b.** Alice creates a change `Record` under her public key worth 0.5 webbBTC. <br />
+   **c.** Alice creates the `Record` to receive 10 webbETH under her public key. This is known as Alice’s receive `Record`. <br />
+   **d.** Bob similarly creates spend, change, and receive records.
+
+<MASPSwapDiagramImage />
+
+1. They then send the following details/proof inputs about these `Record`s to a semi-trusted relayer. The relayer makes the swap ZKP and submits it on-chain. <br /> <br />
+   **a.** These inputs include a signature of the so-called swap message hash `(aliceChangeRecord, aliceReceiveRecord, bobChangeRecord, bobReceiveRecord, t, t')`. This is precisely the data the relayer can tamper with so we have Alice and Bob sign them with their `sk`.
+
+   ```jsx
```

Comment:
Let's improve these docs so there's no circom code. Just the dump of circom inputs/constraints isn't the most fruitful for a reader @saiakilesh